File: Devoirs/DM/DM3/DM3_corrigé.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

struct Graphe{
    int n;
    int** adj;
};

typedef struct Graphe graphe;

// On se contente de trouver la sentinelle -1 et de diviser par deux l'indice
// où elle se trouve.
int degre(graphe G, int s){
    int i = 0;
    // On suppose que le graphe est correctement construit, c'est-à-dire
    // que le tableau G.adj[s][i] contient bien la sentinelle -1.
    while (G.adj[s][i] != -1) i++;
    return i / 2;
}

struct Arete{
    int s;
    int t;
    int p;  
};

typedef struct Arete arete;

void ajouter_arete(graphe G, arete a){
    int degs = degre(G, a.s);
    int degt = degre(G, a.t);
    if (degs == 10 || degt == 10) return;
    // On vérifie que s et t ne sont pas déjà voisins.
    for (int i=0; i<degs; i++){
        if (G.adj[a.s][2 * i] == a.t) return;
    }
    // On modifie alors les tableaux d'adjacence de s et t.
    G.adj[a.s][2 * degs] = a.t;
    G.adj[a.s][2 * degs + 1] = a.p;
    G.adj[a.s][2 * degs + 2] = -1;
    G.adj[a.t][2 * degt] = a.s;
    G.adj[a.t][2 * degt + 1] = a.p;
    G.adj[a.t][2 * degt + 2] = -1;
}

// On ouvre le fichier, on lit la valeur de n, puis on initialise
// le graphe. Ensuite, on lit chaque ligne (tant qu'il y en a) et
// on rajoute l'arête correspondante.
graphe construire_graphe(const char* nom_fichier){
    FILE* canal = fopen(nom_fichier, "r");
    graphe G;
    int success = fscanf(canal, "%d", &(G.n));
    G.adj = malloc(G.n * sizeof(int*));
    for (int s=0; s<G.n; s++){
        G.adj[s] = malloc(21 * sizeof(int));
        G.adj[s][0] = -1;
    }
    while (success != EOF){
        int s, t, p;
        success = fscanf(canal, "%d %d %d", &s, &t, &p);
        if (success == 3){
            arete a = {.s = s, .t = t, .p = p};
            ajouter_arete(G, a);
        }
    }
    fclose(canal);
    return G;
}

// On crée une fonction pour libérer la mémoire occupée par le
// tableau des adjacences d'un graphe (pas besoin de libérer le
// graphe lui-même : il n'a pas d'allocation mémoire).
void liberer_graphe(graphe G){
    for (int s=0; s<G.n; s++){
        free(G.adj[s]);
    }
    free(G.adj);
}

// On somme les poids des arêtes telles que s < t.
int poids_graphe(graphe G){
    int poids = 0;
    for (int s=0; s<G.n; s++){
        for (int i=0; G.adj[s][2 * i] != -1; i++){
            if (s < G.adj[s][2 * i]) poids += G.adj[s][2 * i + 1];
        }
    }
    return poids;
}

// On commence par écrire une fonction de parcours en profondeur récursive,
// de manière similaire à celle écrite en cours.
void DFS(graphe H, int* m, int* cc, int s){
    if (cc[s] != -1) return;
    cc[s] = *m;
    for (int i=0; H.adj[s][2 * i] != -1; i++){
        DFS(H, m, cc, H.adj[s][2 * i]);
    }
}

// Dès lors, on lance un DFS depuis chaque composante connexe, c'est-à-dire
// chaque sommet qui n'a pas encore un numéro de composante.
int* composantes(graphe H, int* m){
    *m = 0;
    int* cc = malloc(H.n * sizeof(int));
    for (int s=0; s<H.n; s++) cc[s] = -1;
    for (int s=0; s<H.n; s++){
        if (cc[s] == -1){
            DFS(H, m, cc, s);
            (*m)++;
        }
    }
    return cc;
}

// Fonction utilitaire qui compare deux arêtes. Renvoie 1 si a > b, -1 si b > a
// et 0 si a = b.
int compare(arete a, arete b){
    if (a.p > b.p) return 1;
    if (a.p < b.p) return -1;
    if (a.s > b.s) return 1;
    if (a.s < b.s) return -1;
    if (a.t > b.t) return 1;
    if (a.t < b.t) return -1;
    return 0;
}

// On parcourt l'ensemble des arêtes de G, et pour chaque arête qui relie deux
// composantes de H, on vérifie pour chacune des deux composantes si elle est
// inférieure à l'arête déjà choisie.
arete* aretes_sures(graphe G, graphe H, int* cc, int m){
    arete* sures = malloc(m * sizeof(arete));
    // On initialise le premier sommet de chaque arete à -1, ce qui correspond
    // au cas où aucune arête n'a encore été choisie.
    for (int i=0; i<m; i++) sures[i].s = -1;
    for (int s=0; s<G.n; s++){
        for (int i=0; G.adj[s][2 * i] != -1; i++){
            int t = G.adj[s][2 * i];
            if (cc[s] == cc[t]) continue;
            arete a = {.s = s, .t = t, .p = G.adj[s][2 * i + 1]};
            if (sures[cc[s]].s == -1 || compare(a, sures[cc[s]]) == -1){
                sures[cc[s]] = a;
            }
            if (sures[cc[t]].s == -1 || compare(a, sures[cc[t]]) == -1){
                sures[cc[t]] = a;
            }
        }
    }
    return sures;
}

// On remarque qu'il existe des arêtes sûres si et seulement si H n'est
// pas connexe. On applique alors l'algorithme tel qu'il est décrit en
// utilisant les fonctions déjà implémentées. On pense bien à libérer
// la mémoire au fur et à mesure.
graphe boruvka(graphe G){
    graphe H;
    H.n = G.n;
    H.adj = malloc(H.n * sizeof(int*));
    for (int s=0; s<H.n; s++){
        H.adj[s] = malloc(21 * sizeof(int));
        H.adj[s][0] = -1;
    }
    int m = H.n;
    while (m != 1){
        int* cc = composantes(H, &m);
        if (m == 1){
            free(cc);
            break;
        }
        arete* sures = aretes_sures(G, H, cc, m);
        for (int i=0; i<m; i++) ajouter_arete(H, sures[i]);
        free(cc);
        free(sures);
    }
    return H;
}

int main(void){
    graphe g_10_ad = construire_graphe("g_10_ad.txt");
    graphe g_10_sd = construire_graphe("g_10_sd.txt");
    graphe g_1000_ad = construire_graphe("g_1000_ad.txt");
    graphe g_1000_sd = construire_graphe("g_1000_sd.txt");
    graphe g_50000_ad = construire_graphe("g_50000_ad.txt");
    graphe g_50000_sd = construire_graphe("g_50000_sd.txt");

    printf("Le degré du sommet 4 dans g_10_ad est %d\n", degre(g_10_ad, 4));

    printf("Le poids du graphe g_10_ad est %d\n", poids_graphe(g_10_ad));
    printf("Le poids du graphe g_10_sd est %d\n", poids_graphe(g_10_sd));
    printf("Le poids du graphe g_1000_ad est %d\n", poids_graphe(g_1000_ad));
    printf("Le poids du graphe g_1000_sd est %d\n", poids_graphe(g_1000_sd));
    // Remarque : il y a un dépassement d'entier pour le poids de g_50000_sd.

    graphe acm_10_ad = boruvka(g_10_ad);
    graphe acm_10_sd = boruvka(g_10_sd);
    graphe acm_1000_ad = boruvka(g_1000_ad);
    graphe acm_1000_sd = boruvka(g_1000_sd);
    graphe acm_50000_ad = boruvka(g_50000_ad);
    graphe acm_50000_sd = boruvka(g_50000_sd);

    printf("Le poids de l'arbre couvrant minimal de g_10_ad est %d\n", poids_graphe(acm_10_ad));
    printf("Le poids de l'arbre couvrant minimal de g_10_sd est %d\n", poids_graphe(acm_10_sd));
    printf("Le poids de l'arbre couvrant minimal de g_1000_ad est %d\n", poids_graphe(acm_1000_ad));
    printf("Le poids de l'arbre couvrant minimal de g_1000_sd est %d\n", poids_graphe(acm_1000_sd));
    printf("Le poids de l'arbre couvrant minimal de g_50000_ad est %d\n", poids_graphe(acm_50000_ad));
    printf("Le poids de l'arbre couvrant minimal de g_50000_sd est %d\n", poids_graphe(acm_50000_sd));

    liberer_graphe(g_10_ad);
    liberer_graphe(g_10_sd);
    liberer_graphe(g_1000_ad);
    liberer_graphe(g_1000_sd);
    liberer_graphe(g_50000_ad);
    liberer_graphe(g_50000_sd);

    liberer_graphe(acm_10_ad);
    liberer_graphe(acm_10_sd);
    liberer_graphe(acm_1000_ad);
    liberer_graphe(acm_1000_sd);
    liberer_graphe(acm_50000_ad);
    liberer_graphe(acm_50000_sd);
    return 0;
}

File: Devoirs/DM/DM3/DM3_corrigé.ml
(* On lance un appel récursif sur la queue et on ajoute
   la tête à la bonne liste en fonction de sa comparaison
   avec le pivot. *)
let rec partition piv = function
    | []     -> [], []
    | x :: q -> let l1, l2 = partition piv q in
                if x <= piv then x :: l1, l2
                            else l1, x :: l2;;

(* Si la liste est de taille 0 ou 1, elle est déjà triée,
   sinon on partition la queue selon la tête, on trie les
   deux moitiés récursivement, et on recolle le tout. *)
let rec tri_rapide lst = match lst with
    | [] | [_] -> lst
    | piv :: q -> let l1, l2 = partition piv q in
                  tri_rapide l1 @ piv :: tri_rapide l2;;

(* Quelques tests : on crée une liste aléatoire, on
   l'affiche, ainsi que sa version triée. On teste
   également sur des listes de taille 0 et 1. *)
let liste_alea n = List.init n (fun _ -> Random.int 20);;

let rec afficher_liste = function
    | []     -> print_newline ()
    | x :: q -> Printf.printf "%d " x; afficher_liste q;;

let _ =
    print_endline "Test du tri rapide.";
    Random.self_init ();
    let lst = liste_alea 20 in
    afficher_liste lst;
    afficher_liste (tri_rapide lst);
    afficher_liste (tri_rapide []);
    afficher_liste (tri_rapide [10]);;

(* On commence par ouvrir le canal et lire la taille du graphe. 
   On crée le graphe, puis on lit chaque ligne du fichier. On 
   crée l'arête correspondante à chaque fois. On gère la levée
   d'exception avec try … with. *)

let construire_graphe nom_fichier =
    let canal = open_in nom_fichier in
    let ios = int_of_string in
    let n = ios (input_line canal) in
    let g = Array.make n [] in
    try while true do
        let s, t, p = match String.split_on_char ' ' (input_line canal) with
           | [s'; t'; p'] -> ios s', ios t', ios p'
           | _            -> failwith "Mauvais format" in
        g.(s) <- (t, p) :: g.(s);
        g.(t) <- (s, p) :: g.(t)
    done;
    g
    with End_of_file -> close_in canal; g;;

(* On construit les différents graphes et on crée une liste les contenant. *)

let g_10_ad = construire_graphe "g_10_ad.txt";;
let g_10_sd = construire_graphe "g_10_sd.txt";;
let g_1000_ad = construire_graphe "g_1000_ad.txt";;
let g_1000_sd = construire_graphe "g_1000_sd.txt";;
let g_50000_ad = construire_graphe "g_50000_ad.txt";;
let g_50000_sd = construire_graphe "g_50000_sd.txt";;

let graphes = [g_10_ad; g_10_sd; g_1000_ad; g_1000_sd; g_50000_ad; g_50000_sd];;

(* On utilise une référence et on parcourt toutes les listes d'adjacence. On
   n'ajoute le poids d'une arête que si s < t, pour éviter de les compter 
   deux fois (ce qu'on aurait pu faire en divisant par deux à la fin). *)

let poids_graphe g =
    let n = Array.length g in
    let poids = ref 0 in
    for s = 0 to n - 1 do
        let aux (t, p) =
            if s < t then poids := !poids + p in
        List.iter aux g.(s)
    done;
    !poids;;

let _ =
    let aux i g = 
        Printf.printf "Le poids du graphe d'indice %d est %d\n" i (poids_graphe g) in
    List.iteri aux graphes;;

(* Une partition de taille n est simplement un tableau de -1, car 
   les singletons sont de rang nul. *)
let creer_partition n = Array.make n (-1);;

(* Comme la fonction du cours (mais en OCaml). *)
let rec trouver partition x =
    if partition.(x) < 0 then x
    else begin
       let rx = trouver partition partition.(x) in
       partition.(x) <- rx;
       rx
    end;;

(* Similaire à la fonction du cours, sauf qu'on traite le cas
   d'égalité différemment. On garde bien en mémoire que si 
   partition.(rx) < partition.(ry), alors le rang de rx est 
   SUPÉRIEUR à celui de ry (à cause des valeurs négatives). *)
let unir partition x y =
    let rx = trouver partition x and ry = trouver partition y in
    if rx <> ry then begin
        if partition.(rx) < partition.(ry) then
            partition.(ry) <- rx
        else if partition.(rx) > partition.(ry) then
            partition.(rx) <- ry
        else begin
            partition.(rx) <- partition.(rx) - 1;
            partition.(ry) <- rx
        end
    end;;

(* Similaire à la fonction poids_graphe. On trie la liste une fois
   qu'elle est créée (on utilise l'ordre lexicographique comme ordre
   naturel, c'est pour ça qu'on met les poids en premier). *)
let aretes g =
    let n = Array.length g in
    let lst_aretes = ref [] in
    for s = 0 to n - 1 do
        let aux (t, p) =
            if s < t then lst_aretes := (p, s, t) :: !lst_aretes in
        List.iter aux g.(s)
    done;
    tri_rapide !lst_aretes;;

(* Un petit test. *)
let _ =
    let lst_aretes = aretes g_10_sd in
    let p, s, t = List.hd lst_aretes in
    Printf.printf "L'arête de poids minimal de g_10_sd est (%d, %d), de poids %d\n" s t p;
    let p', s', t' = List.hd (List.rev lst_aretes) in
    Printf.printf "L'arête de poids maximal de g_10_sd est (%d, %d), de poids %d\n" s' t' p';;

(* Après avoir créé un graphe pour l'arbre couvrant et la partition, on
   crée une fonction auxiliaire qui permet de traiter individuellement
   une arête (la rajouter à l'arbre et fusionner les deux classes ou ne
   rien faire). On l'applique à chaque arête dans l'ordre donné par la
   fonction aretes. *)
let kruskal g =
    let n = Array.length g in
    let acm = Array.make n [] in
    let partition = creer_partition n in
    let traiter_arete (p, s, t) =
        let rs = trouver partition s and rt = trouver partition t in
        if rs <> rt then begin
           acm.(s) <- (t, p) :: acm.(s);
           acm.(t) <- (s, p) :: acm.(t);
           unir partition s t
        end in
    List.iter traiter_arete (aretes g);
    acm;;

(* On calcule les poids des ACM. *)
let _ =
    let aux i g = 
        Printf.printf "Le poids de l'arbre couvrant minimal du graphe d'indice %d est %d\n" 
                      i (poids_graphe (kruskal g)) in
    List.iteri aux graphes;;